source/MaterialXGenShader/Util.cpp

```diff
@@ -839,4 +841,44 @@ void mapNodeDefToIdentiers(const std::vector<NodePtr>& nodes,
     }
 }
 
+void findLights(DocumentPtr doc, std::vector<NodePtr>& lights)
+{
+    lights.clear();
+    for (NodePtr node : doc->getNodes())
+    {
+        const TypeDesc* type = TypeDesc::get(node->getType());
+        if (type == Type::LIGHTSHADER)
+        {
+            lights.push_back(node);
+        }
+    }
+}
+
+void registerLights(DocumentPtr doc, const std::vector<NodePtr>& lights, GenContext& context)
+{
+    // Clear context light user data which is set when bindLightShader() 
+    // is called. This is necessary in case the light types have already been
+    // registered.
+    context.popUserData(HW::USER_DATA_LIGHT_SHADERS);
+
+    if (!lights.empty())
+    {
+        // Create a list of unique nodedefs and ids for them
+        std::unordered_map<std::string, unsigned int> identifiers;
+        mapNodeDefToIdentiers(lights, identifiers);
```

Comment:
The ids bound to light shaders here doesn't seem to be returned.. How do you know what ids to use when assigning light data to the generated shader later?

Reply:
For now this needs to be called more than once. At actual shader bind time we recall this. I guess it would make sense to allow this to be returned if the user wanted to keep this persistent.